src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PathSeqFilterSpark.java

```diff
@@ -1,268 +1,115 @@
-
-        //Load Kmer hopscotch set and filter reads containing > 0 matching kmers
-        final JavaRDD<GATKRead> readsKmerFiltered = doKmerFiltering(ctx,readsAmbigFiltered);
-        logger.info("Reads remaining after kmer filtering: " + readsKmerFiltered.count());
-
-        //Filter unpaired reads
-        final JavaRDD<GATKRead> readsFilteredPaired = retainPairs(readsKmerFiltered);
-        logger.info("Reads remaining after unpaired filtering: " + readsFilteredPaired.count());
-
-        //BWA filtering against user-specified host organism reference
-        header = getHeaderForReads();
-        final JavaRDD<GATKRead> readsAligned = doHostBWA(ctx, header, readsFilteredPaired);
-
-        //Get unmapped reads (note these always come in pairs)
-        //TODO: retain read pairs by alignment score instead of flags
-        final JavaRDD<GATKRead> readsNonHost = readsAligned.filter(read -> read.isUnmapped() && read.mateIsUnmapped());
-        logger.info("Reads remaining after BWA filtering: " + readsFilteredPaired.count());
-
-        //TODO: repeat BWA with seed size 11
-
-        //Write output
-        header.setSortOrder(SAMFileHeader.SortOrder.queryname);
-        try {
-            ReadsSparkSink.writeReads(ctx, OUTPUT_PATH, null, readsNonHost, header, shardedOutput ? ReadsWriteFormat.SHARDED : ReadsWriteFormat.SINGLE);
-        } catch (final IOException e) {
-            throw new GATKException("Unable to write bam",e);
+        final SAMFileHeader header = getHeaderForReads();
+        if (filterArgs.alignedInput && (header.getSequenceDictionary() == null || header.getSequenceDictionary().isEmpty())) {
+            logger.warn("--isHostAligned is true but the BAM header contains no sequences");
+        }
+        if (!filterArgs.alignedInput && header.getSequenceDictionary() != null && !header.getSequenceDictionary().isEmpty()) {
+            logger.warn("--isHostAligned is false but there are one or more sequences in the BAM header");
+        }
+        header.setSequenceDictionary(new SAMSequenceDictionary());
+
+        try (final PSFilter filter = new PSFilter(ctx, filterArgs, getReads(), header)) {
+
+            final Tuple2<JavaRDD<GATKRead>, JavaRDD<GATKRead>> result = filter.doFilter();
+            final JavaRDD<GATKRead> pairedReads = result._1;
+            final JavaRDD<GATKRead> unpairedReads = result._2;
+
+            if (!pairedReads.isEmpty()) {
+                final SAMFileHeader.SortOrder oldSortOrder = header.getSortOrder();
+                header.setSortOrder(SAMFileHeader.SortOrder.queryname);
+                writeReads(ctx, outputPath + ".paired.bam", pairedReads, header);
+                header.setSortOrder(oldSortOrder);
+            } else {
+                logger.info("No paired reads to write - BAM will not be written.");
+            }
+            if (!unpairedReads.isEmpty()) {
+                writeReads(ctx, outputPath + ".unpaired.bam", unpairedReads, header);
```

Comment:
Don't you need to dink around with the sort order specifier in the header, just like for the paired case?